### PR TITLE
Refactor Journey class to be easier to understand and extend

### PR DESCRIPTION
**Pull request by lfdebrux**
### What problem does this pull request solve?

Trello card: https://trello.com/c/mOvI9EEw/43-allow-form-preview-to-skip-to-a-specific-question 

As part of firebreak, I ended up having to play around with the `Journey` class that we use to hold the steps that a form filler has followed while filling out a form.

I wanted to create a subclass that tweaked the behaviour slightly, but I found this class a bit hard to wrap my head around, with a lot of tight logic wrapped up into one method.

This PR does a big refactor with the aim of making the class easier to understand (at least, I think it is easier to understand now). I also added some comments trying to explain what I think the class is for. I also added a bunch more tests, so I could feel confident that the refactor retained the essential behaviour.

### Things to consider when reviewing

- Ensure that you consider the wider context.
- Does it work when run on your machine?
- Is it clear what the code is doing?
- Do the commit messages explain why the changes were made?
- Are there all the unit tests needed?
- Has all relevant documentation been updated

### CODE REVIEW

**Review by stephencdaly — COMMENTED** [review]
I think these changes are really good and really helpful.

I struggled a little with the last commit that refactored `journey.rb` - I've left a question and a suggestion. You might need to talk me through some of the changes.

File: spec/lib/flow/journey_spec.rb [hunk 1]
```diff
@@ -55,6 +71,98 @@
...
+
+        context "and the repeatable question has been answered more than once" do
+          let(:store) { { answers: { "2" => { "1" => { selection: "Option 2" }, "2" => [{ text: "Example text" }, { text: "Different example text" }], "3" => { text: "More example text" } } } } }
+
+          it "includes all pages once each" do
+            expect(journey.completed_steps.to_json).to eq [first_step_in_journey, second_step_in_journey, third_step_in_journey].to_json
+          end
+        end
+
+        context "but the answer store has answers for a non-repeatable question" do
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply spec/lib/flow/journey_spec.rb:142]: Yes, that's exactly it. That can happen when a form creator is previewing a form while making changes to it, and changes a question to be repeatable after they have already answered it in the preview.

File: app/lib/flow/journey.rb [hunk 2]
```diff
@@ -24,37 +24,58 @@ class Journey
...
+          break if next_page_slug.nil?
+
+          current_step = @step_factory.create_step(next_page_slug)
+        end
+      end
+    end
 
-        @completed_steps << current_step
+    def safe_load_from_context(step)
+      return step unless step.respond_to? :load_from_context
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply app/lib/flow/journey.rb:71]: `respond_to?` tests whether an object has a method with that name: see https://www.rubydoc.info/stdlib/core/Object#respond_to%3F-instance_method.

In the original code it could be assumed that the step had a `load_from_context` method because the load only happened if `@form_context.get_stored_answer(step)` was true, and so if you have answer data in the context for a step then it makes sense that the step should be able to load that data from the context... but if for instance the step is a `CheckYourAnswersStep` that shouldn't have anything in the answer store and it's fine that it doesn't implement the `load_from_context` method.

That said it took me a little while to work that logic out. I personally feel that it's better to explicitly check that the step can load from the context, which is what `respond_to?` does.

File: app/lib/flow/journey.rb [hunk 3]
```diff
@@ -24,37 +24,58 @@ class Journey
...
-      @completed_steps = []
-      current_step = find_completed_step(:_start)
+      each_step_with_routing.take_while do |step|
+        step_is_completed?(step)
+      end
+    end
+
+    def each_step_with_routing
+      current_step = @step_factory.create_step(:_start)
+      loop_detection = []
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply app/lib/flow/journey.rb:45]: I wanted the name to describe what the variable was for, but I'm happy to change it if it's confusing.

File: app/lib/flow/journey.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephencdaly) [reply app/lib/flow/journey.rb:69]: Oh so the step might not actually be a Step?! 😨 

Ok I guess this makes sense then. Reckon it's worth adding a comment to that effect, as it's a little hard to follow what this is doing unless you fully follow through from where we load the steps.

Why do we need to duplicate the step into `original_step`?

[the remark this answers is not in the document]
Comment (lfdebrux) [reply app/lib/flow/journey.rb:69]: > Why do we need to duplicate the step into original_step?

It's just a precaution really, but the `load_from_context` might not be atomic, if step is a `RepeatableStep` it could fail halfway through and then step would have some but not all of the data loaded into it.

File: spec/lib/flow/journey_spec.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephencdaly) [reply spec/lib/flow/journey_spec.rb:142]: Is there a way to reword that slightly to make it clearer? How about:

"but the answer store does not have answers in the format expected for the repeatable question"

[the remark this answers is not in the document]
Comment (lfdebrux) [reply spec/lib/flow/journey_spec.rb]: I've changed it to

https://github.com/alphagov/forms-runner/blob/3810e592e14ec66cf1f756b5a774f5e1f62fee76/spec/lib/flow/journey_spec.rb#L142

File: app/lib/flow/journey.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (lfdebrux) [reply app/lib/flow/journey.rb]: I've added some more comments:

https://github.com/alphagov/forms-runner/blob/3810e592e14ec66cf1f756b5a774f5e1f62fee76/app/lib/flow/journey.rb#L68-L71

[the remark this answers is not in the document]
Comment (stephencdaly) [reply app/lib/flow/journey.rb]: Thanks!
